pkg/common/cns-lib/volume/taskmanager.go

```diff
@@ -0,0 +1,356 @@
+				for _, prop := range update.ChangeSet {
+					log.Debugf("Got update for object %v properties %v", update.Obj, prop)
+					if prop.Name == infoPropertyName {
+						go t.handleTaskUpdate(ctx, prop)
+					}
+				}
+			}
+
+			return t.shouldStopListening
+		})
+		if err != nil {
+			log.Errorf("WaitForUpdates returned err: %v, vimClient: %v", err, t.virtualCenter.Client.Client.UserAgent)
+			if soap.IsSoapFault(err) {
+				switch soap.ToSoapFault(err).VimFault().(type) {
+				case *types.ManagedObjectNotFound, types.ManagedObjectNotFound:
+					recreateView = true
+				case *types.NotAuthenticated, types.NotAuthenticated:
+					log.Info("observed NotAuthenticated error. should recreate view\n")
+					recreateView = true
+					if err := t.virtualCenter.Connect(ctx); err != nil {
+						log.Errorf("failed to connect to vCenter. err: %v", err)
+					} else {
+						log.Infof("connection to vc successful")
+					}
+				}
+			}
+			// if any error is received from vpxd, we immediately return the fault for all the pending tasks in the map
+			// this is consistent with the current implementation
+			for taskMoRef := range t.taskMap {
+				prop := types.PropertyChange{
+					Val: types.TaskInfo{
+						Task:  taskMoRef,
+						State: types.TaskInfoStateError,
+						Error: &types.LocalizedMethodFault{
+							LocalizedMessage: err.Error(),
+						},
+					},
+				}
+				go t.handleTaskUpdate(ctx, prop)
+			}
+			time.Sleep(waitForUpdatesRetry)
+		}
+	}
+}
+
+// handleTaskUpdate is processes each task update in a separate goroutine
+func (t *TaskViewManagerImpl) handleTaskUpdate(ctx context.Context, prop types.PropertyChange) {
+	log := logger.GetLogger(ctx)
+	log.Debugf("handleTaskUpdate for prop: %+v", prop)
+	taskInfo, ok := prop.Val.(types.TaskInfo)
+	if !ok {
+		log.Errorf("failed to cast taskInfo for prop: %+v", prop)
+		return
+	}
+	if taskInfo.State == types.TaskInfoStateQueued || taskInfo.State == types.TaskInfoStateRunning {
+		log.Infof("task not completed yet. current state: %v", taskInfo.State)
+		return
+	}
+	result := Result{}
+	t.mu.Lock()
+	requestorChan, err := t.getChannelFromTask(ctx, taskInfo.Task)
+	t.mu.Unlock()
+	if err != nil {
+		result.TaskInfo = nil
+		result.Err = err
+	} else if taskInfo.State == types.TaskInfoStateError {
+		log.Debugf("taskInfo.Error: %+v", taskInfo.Error)
+		result.TaskInfo = nil
+		result.Err = errors.New(taskInfo.Error.LocalizedMessage)
+	} else {
+		result.TaskInfo = &taskInfo
+		result.Err = nil
+	}
+
+	requestorChan <- result
+}
+
+// AddTaskToListView adds task to listView and the internal map
+func (t *TaskViewManagerImpl) AddTaskToListView(ctx context.Context, taskMoRef types.ManagedObjectReference, ch chan Result) error {
+	log := logger.GetLogger(ctx)
+	log.Debugf("AddTaskToListView for %+v", taskMoRef)
+	for t.getListView() == nil {
+		log.Debugf("listview not created. waiting for 0.5 sec")
+		time.Sleep(time.Millisecond * 500)
+	}
+	listView := t.getListView()
```

Comment:
`t.listview` is directly accessible in the `AddTaskToListView`. We do not need `t.getListView()`

Reply:
Fixed.